### ISSUE TITLE
Proposed logging solution

### ISSUE THREAD

**Issue by mmcev106** [vanderbilt/redcap-external-modules#87]
I am planning to add basic logging support for a project I'm currently working on.  Several people have expressed interest in this feature, so I'm looking for input to try and make sure what I'm building is as useful as possible.  I’ve been brainstorming on what the API for such a feature would look like, and came up with the log() and queryLog() methods documented here:

https://github.com/vanderbilt/redcap-external-modules/blob/logging/docs/official-documentation.md

The general SQL-like syntax for queryLog() seems ambitious, but I've actually already implemented that for a WordPress plugin that's been in production and working like a charm for eight months.  We could also support a more basic array syntax if need be.  The specifics of the user and timestamp are a little simplistic, so those would likely get flushed out during implementation.  I’m envisioning two new database tables to support this (something like **redcap_external_module_log_entry** and **redcap_external_module_log_entry_parameter**).

What does everyone think?  Is this a nice and flexible general solution?  Do you see any limitations or room for improvement upfront?

@123andy, @tbembersimeao, @pbchase

**Comment by 123andy** [comment]
Are you intending for this to be for storing record-level metadata related to an external module?  Such as whether or not record 55 has been sent email template 3 for event 123 ?

-A

—
Andrew Martin, PhD
Stanford University School of Medicine

On Mar 1, 2018, at 8:48 AM, Mark McEver <notifications@github.com<mailto:notifications@github.com>> wrote:


I am planning to add basic logging support for a project I'm currently working on. Several people have expressed interest in this feature, so I'm looking for input to try and make sure what I'm building is as useful as possible. I’ve been brainstorming on what the API for such a feature would look like, and came up with the log() and queryLog() methods documented here:

https://github.com/vanderbilt/redcap-external-modules/blob/logging/docs/official-documentation.md

The general SQL-like syntax for queryLog() seems ambitious, but I've actually already implemented that for a WordPress plugin that's been in production and working like a charm for eight months. We could also support a more basic array syntax if need be. The specifics of the user and timestamp are a little simplistic, so those would likely get flushed out during implementation. I’m envisioning two new database tables to support this (something like redcap_external_module_log_entry and redcap_external_module_log_entry_parameter).

What does everyone think? Is this a nice and flexible general solution? Do you see any limitations or room for improvement upfront?

@123andy<https://github.com/123andy>, @tbembersimeao<https://github.com/tbembersimeao>, @pbchase<https://github.com/pbchase>

**Comment by mmcev106** [comment]
Now that you mention it, I'm thinking it makes sense for record, event, instrument, and instance to always be stored automatically like user and timestamp.

**Comment by mmcev106** [comment]
We are planning on moving forward with this implementation using the following table structure:

```
CREATE TABLE `redcap_external_modules_log` (
 `log_id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
 `created_at` datetime DEFAULT NULL,
 `ui_id` int(11) DEFAULT NULL,
 `external_module_id` int(11) DEFAULT NULL,
 `project_id` int(11) DEFAULT NULL,
 `record` varchar(100) DEFAULT NULL,
 `message` mediumtext,
 PRIMARY KEY (`log_id`),
 KEY `redcap_log_created_at_idx` (`created_at`),
 KEY `redcap_log_record_idx` (`record`),
 KEY `redcap_log_message_idx` (`message`(255)),
 KEY `redcap_log_redcap_projects_record` (`project_id`,`record`) USING BTREE,
 KEY `redcap_log_redcap_external_modules_fk` (`external_module_id`),
 KEY `redcap_log_redcap_user_information_fk` (`ui_id`),
 CONSTRAINT `redcap_log_redcap_external_modules_fk` FOREIGN KEY (`external_module_id`) REFERENCES `redcap_external_modules` (`external_module_id`) ON DELETE CASCADE ON UPDATE CASCADE,
 CONSTRAINT `redcap_log_redcap_projects_fk` FOREIGN KEY (`project_id`) REFERENCES `redcap_projects` (`project_id`) ON DELETE CASCADE ON UPDATE CASCADE,
 CONSTRAINT `redcap_log_redcap_user_information_fk` FOREIGN KEY (`ui_id`) REFERENCES `redcap_user_information` (`ui_id`) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8;

CREATE TABLE `redcap_external_modules_log_parameters` (
 `log_id` bigint(20) unsigned NOT NULL,
 `name` varchar(255) NOT NULL,
 `value` mediumtext,
 PRIMARY KEY (`log_id`,`name`),
 KEY `redcap_log_parameters_name_idx` (`name`),
 KEY `redcap_log_parameters_value_idx` (`value`(255)),
 CONSTRAINT `redcap_log_parameters_redcap_log_fk` FOREIGN KEY (`log_id`) REFERENCES `redcap_external_modules_log` (`log_id`) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

@123andy, @tbembersimeao, @pbchase, do you foresee any logging needs that this structure would not support?

**Comment by 123andy** [comment]
Looks interesting.  I think that could work well.  Are you thinking of some helper methods for the ExternalModule class too?

**Comment by mmcev106** [comment]
For my current project, I think the only methods I'll need are the two mentioned in the first comment above.  Did you have any others in mind?

**Comment by 123andy** [comment]
Hey @mmcev106  - playing with the queryLogs...  Really a cool idea!

Getting some odd behavior with a few queries:

"select *" works as expected.
"select *, foo" where foo is a key that was stored but it leaves the 'name' and 'value' keys in the result array.  Not sure if this should happen.
```
Array
(
    [log_id] => 2
    [timestamp] => 2018-12-05 15:34:52
    [ui_id] => 3
    [ip] => 172.18.0.1
    [external_module_id] => 21
    [project_id] => 21
    [record] => 
    [message] => test message - with params
    [name] => foo
    [value] => bar
    [foo] => bar
)
```

This gets more confusing when you try
"select *, foo, foo2" since the 'last' values for 'name' and 'value' are foo2.


Second oddity - I tried storing a key called 'user-record'.  Goes in fine and I can see it in the settings table, but when I try to query for it I get a function or something like that?
```
Array
(
    [user() - redcap_external_modules_log.record] => 
)
```


And, lastly, another question.  You currently don't allow overriding any of the 'reserved' keys.  I can envision, however, use cases where you might want to override the record id.  Say you have a cron task that is sending out notifications to records and then recording to the log that a notification was sent out for that record.    It seems a shame to leave the main 'record' value blank and have to make up a different name for the record_id as a key/value pair in the parameters array...  Maybe there should be an exemption that allows override of record in the parameter array...

**Comment by mmcev106** [comment]
Using "select *" is actually not supported, and looks like it causes some odd behavior (the name & value columns).  I just made a change to throw an exception if it is attempted to clarify this.

We did actually decide to allow overriding "record" via PHP (though it is not via JS) per [this discussion](https://community.projectredcap.org/questions/46311/external-modules-logging-framework.html), although it came a bit later (maybe you're not on that version yet).

I'm not seeing the odd "user-record" behavior...  Perhaps I'm trying to reproduce it a little differently, or maybe it fixed at some point as well.

I'd love to hear any and all input you might have on this feature.  I just removed the "BETA" flags from the docs a few days ago.  It's working quite well for my analytics modules, but I have not had any other use cases yet, and haven't had much feedback from others.
